### PR TITLE
braintree initial support with transactions/disputes

### PR DESCRIPTION
**Pull request by schuyler1d**
TODO:
 - [x] documentation
 - [x] tests
 - [x] adding some more logging

### CODE REVIEW

**Comment by jburchard** [comment]
Is this ready for review? Just want to make sure that we aren't holding you up on our end.

**Comment by schuyler1d** [comment]
Yes, this should be ready for review (along with all my current open PRs).

File: parsons/braintree/braintree.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+            Not required if ``BRAINTREE_MERCHANT_ID`` env variable is set.
+        public_key: str
+            Braintree public key -- probably a (different) 16-char alphanumeric.
+            Not required if ``BRAINTREE_PUBLIC_KEY`` env variable is set.
+        private_key: str
+            Braintree private key -- probably a 32-char alphanumeric.
+            Not required if ``BRAINTREE_PRIVATE_KEY`` env variable is set.
+        timeout: int
+            Optionally change the timeout from default of 200 seconds.
+            Can also be passed with env var ``BRAINTREE_TIMEOUT``.
```
Comment (jburchard) [inline parsons/braintree/braintree.py:30]: Missing the `production` kwarg.

File: parsons/braintree/braintree.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+        timeout = check_env('BRAINTREE_TIMEOUT', timeout, optional=True) or 200
+
+        self.gateway = braintree.BraintreeGateway(
+            braintree.Configuration(
+                environment=(braintree.Environment.Production if production
+                             else braintree.Environment.Sandbox),
+                merchant_id=merchant_id,
+                public_key=public_key,
+                private_key=private_key,
+                timeout=200))
```
Comment (jburchard) [inline parsons/braintree/braintree.py:147]: I think it should be `timeout=timeout`.

File: parsons/braintree/braintree.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+        """
+        Get a table of disputes based on query parameters.
+        There are three ways to pass query arguments: Pass a start_date and end_date
+        together for a date range, or pass a query_list or query_dict argument.
+
+        `Args:`
+            start_date: datetime or str
+                Start date of the dispute range. Requires `end_date` arg.
+            end_date: datetime or str
+                End date of the dispute range. Requires `start_date` arg.
```
Comment (jburchard) [inline parsons/braintree/braintree.py:162]: Might want to specify the format of the datas, iso etc and provide an example.

File: parsons/braintree/braintree.py [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+            query_list: list of braintree.DisputeSearch
+                You can use the `braintree.DisputeSearch
+                <https://developers.braintreepayments.com/reference/request/dispute/search/python>`
```
Comment (jburchard) [inline parsons/braintree/braintree.py:165]: Change to markdown linking, e.g. ``` `API documentation <https://developers.ngpvan.com/van-api#van>`_ ```

File: parsons/braintree/braintree.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+        for a date range, or pass a query_list or query_dict argument.
+
+        `Args:`
+            disbursement_start_date: datetime or str
+                Start date of the disbursement range. Requires `disbursement_end_date` arg.
+            disbursement_end_date: datetime or str
+                End date of the disbursement range. Requires `disbursement_start_date` arg.
+            query_list: list of braintree.DisputeSearch
+                You can use the `braintree.DisputeSearch
+                <https://developers.braintreepayments.com/reference/request/dispute/search/python>`
```
Comment (jburchard) [inline parsons/braintree/braintree.py:215]: Change to markdown linking, e.g. ``` `API documentation <https://developers.ngpvan.com/van-api#van>`_ ```

File: parsons/braintree/braintree.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+               to create a manual list of query parameters.
+            query_dict: jsonable-dict
+                query_dict is basically the same as query_list, except instead of using their API
+                objects, you can pass it in pure dictionary form.
+                Some examples:
+                    The disbursement_start_date/dibursement_end_date arguments are the same as
+                      `{"disbursement_date": {"between": [start_date, end_date]}}`
+                    `{"merchant_account_id": {"in_list": [123, 456]}}`
+                    `{created_at: {"greater_than_or_equal": "2020-03-10"}}`
```
Comment (jburchard) [inline parsons/braintree/braintree.py:224]: More weird markdowness. I'd recommend multi-line like this.

Examples:
```
.. highlight:: python
.. code-block:: python
  [{'score1_id' : int, score1_column': str}
   {'score2_id' : int, score2_column': str}]
```

**Comment by schuyler1d** [comment]
thanks for the careful review -- I think I finally am getting the hang of sphinx (again).

**Comment by schuyler1d** [comment]
I'm not sure what the dockercloud error is -- could someone report what the issue is?

**Comment by jburchard** [comment]
Looks like it passed the tests, just took a few minutes.

**Pull request merged**